### ISSUE TITLE
crashes in minecraft server causing loss of components

### ISSUE THREAD

**Issue by payonel** [MightyPirates/OpenComputers#1991]
In some areas (issue currently in btm world) we are losing components, or they become "stale" blocks in the world

In a computer case, all of the internals are lost, as if removed
In a server rack, the blades are lost, as if removed
cables, keyboards, and power converted (possible other blocks) become "stale" -- they DO NOT post as a component when a computer is placed next to them. They have to be destroyed (creative hit) and replaced in the world.

It appears that a server crash is the catalyst for the repro.

It doesn't affect all computers in the world, and it **DOES** appear to be chunk specific.

I will ping unascribed to kindly provide us with the crash report.

We know we had a crash and a repro of this bug at approx 20:03:55 BST

**Comment by payonel** [comment]
[18:34:48] [Server thread/WARN] [OpenComputers]: A component of type 'internet' disappeared (71d31312-2277-4634-a1ae-cf996ceb2de8)! This usually means that it didn't save its node.

**Comment by exaskye** [comment]
I can't find any exceptions from when this happens, but after the server restarts, I get a number of lines like:

```
[18:34:48] [Server thread/WARN] [OpenComputers/]: A component of type 'internet' disappeared (71d31312-2277-4634-a1ae-cf996ceb2de8)! This usually means that it didn't save its node.
[19:12:15] [Server thread/WARN] [OpenComputers/]: A component of type 'screen' disappeared (56c4c2c6-23f6-4ba7-9e4a-74ca47dcdcdf)! This usually means that it didn't save its node.
[19:12:15] [Server thread/WARN] [OpenComputers/]: A component of type 'chat_box' disappeared (fa3ba52a-5f3c-43d6-8d73-7135a3d2d3a9)! This usually means that it didn't save its node.
[19:12:15] [Server thread/WARN] [OpenComputers/]: A component of type 'keyboard' disappeared (466b4606-854a-42f6-8786-77b64215ce03)! This usually means that it didn't save its node.
```

**Comment by fnuecke** [comment]
Those warnings are to be expected and perfectly normal if components die in a save (since the machine state knew them before the save, but they're gone after).

If the crash is not related to OC, then I'm afraid there's not really anything I see I can do to prevent this. If it _is_ OC, I'd need some logs of the actual crash.

As for "stale" blocks, I'm guessing their NBT might be semi-wiped, with their persisted node address being an empty string or something... I'll add a check on load to ignore persisted empty strings, maybe that'll at least allow those to recover somewhat without having to be replaced.

**Comment by exaskye** [comment]
The server was stopped gracefully, and I can't find any OC errors in the log before it stops.

**Comment by fnuecke** [comment]
Well damn. I can't reproduce any loss of data in dev (running a server with a bunch of stuff in a few chunks, stopping and restarting it a few times). No clue how to try and track this down then, I'm afraid :/

**Comment by payonel** [comment]
btw, i'm now 95% sure this is chunk specific

**Comment by inklit** [comment]
This is the same behavior that happens after #1990, messages about components disappearing included.

**Comment by brisingraerowing** [comment]
I am having the same issue in a singleplayer world, and I also noticed that sometimes the computer case will turn black and will have at least one 'ghost' component in it (usually a CPU or APU), and won't be able to be used until broken and replaced.

I'll try and get screenshots and logs as soon as I can.

**Comment by magik6k** [comment]
Sounds like #2092, try updating to newest build.

**Comment by brisingraerowing** [comment]
@magik6k I am using a build I made from git about 3 weeks ago, although I will try and do a new build.

**Comment by brisingraerowing** [comment]
I still get this with the latest 1.10.2 git code. Cases turning black and components going missing. No cables or conduits either. All blocks are just touching each other. This is in singleplayer after a world reload.

**Comment by brisingraerowing** [comment]
[Here is a log](https://gist.github.com/BrisingrAerowing/3e1f95def707272166e7e1d4a545d6ce)

It looks like there is an ArrayIndexOutOfBounds exception as the root cause of the errors (found near the bottom)

@fnuecke You might want to take a look at that.

**Comment by Vexatos** [comment]
I do not see anything related to OC in that log.

**Comment by brisingraerowing** [comment]
```
Caused by: java.lang.ArrayIndexOutOfBoundsException: 8
	at li.cil.oc.common.inventory.ComponentInventory$class.onItemRemoved(ComponentInventory.scala:142) ~[ComponentInventory$class.class:?]
	at li.cil.oc.common.tileentity.Case.li$cil$oc$common$tileentity$traits$ComponentInventory$$super$onItemRemoved(Case.scala:28) ~[Case.class:?]
	at li.cil.oc.common.tileentity.traits.ComponentInventory$class.onItemRemoved(ComponentInventory.scala:100) ~[ComponentInventory$class.class:?]
	at li.cil.oc.common.tileentity.Case.onItemRemoved(Case.scala:109) ~[Case.class:?]
	at li.cil.oc.common.inventory.Inventory$class.setInventorySlotContents(Inventory.scala:32) ~[Inventory$class.class:?]
	at li.cil.oc.common.tileentity.Case.func_70299_a(Case.scala:28) ~[Case.class:?]
	at li.cil.oc.common.inventory.SimpleInventory$class.decrStackSize(SimpleInventory.scala:27) ~[SimpleInventory$class.class:?]
	at li.cil.oc.common.tileentity.Case.func_70298_a(Case.scala:28) ~[Case.class:?]
```

You have to view raw to see it.

**Comment by brisingraerowing** [comment]
Anyone? @fnuecke 
@Vexatos
?

**Comment by payonel** [comment]
i found that some mods caused the component inventory to load incomplete (tesla or something). I fixed this some builds back by delaying the inventory size check until really needed. One of the side affects is that it would cause the inventory size to be too small, and in that stack trace given about onItemRemoved, there would be an array out of band exception.
So the early-load issue has been fixed, and I'll add another length check for safety
